src/persist/src/indexed/columnar.rs

```diff
@@ -13,22 +13,48 @@
 use std::fmt;
 use std::iter::FromIterator;
 
-use serde::{Deserialize, Serialize};
+use arrow2::buffer::{Buffer, MutableBuffer};
+use arrow2::types::Index;
+use ore::cast::CastFrom;
 
-/// A set of ((Key, Val), Time, Diff) records stored in a columnar representation.
+pub mod arrow;
+pub mod parquet;
+
+/// A set of ((Key, Val), Time, Diff) records stored in a columnar
+/// representation.
+///
+/// Note that the data are unsorted, and unconsolidated (so there may be
+/// multiple instances of the same ((Key, Val), Time), and some Diffs might be
+/// zero, or add up to zero).
+///
+/// The i'th key's data is stored in
+/// `key_data[key_offsets[i]..key_offsets[i+1]]`. Similarly for val.
+///
+/// Invariants:
+/// - key_offsets.len() == len + 1
```

Comment:
Done (you meant <= right?). I didn't add a check for this to validate since it seemed silly

Reply:
no I meant <, because `key_offsets.len()` has to be able to fit in a `usize`, and has to be equal to `len + 1` so len can't be `usize::MAX` and has to be less than that iiuc.

Reply:
Ah! That does make more sense. Will do